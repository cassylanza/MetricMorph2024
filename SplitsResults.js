import React, { Component } from 'react'
import { useState } from 'react'
import { TouchableOpacity, Text, View, StyleSheet, ScrollView, FlatList } from 'react-native'
import { Actions } from 'react-native-router-flux'

// formats the calculated time into hours, minutes, and seconds
// returns default of "hh:mm:ss" if input is invalid (i.e. not a number
// or outside of the specified range)
export function formatResult(timeInSeconds) {

  let maxNumber = 362399; // formats to 99:59:59

  if (isNaN(timeInSeconds) || typeof timeInSeconds !== 'number') {
      return "hh:mm:ss";
  }

  if(timeInSeconds < 0 || timeInSeconds > maxNumber) {
    return "hh:mm:ss";
  }

  let hours = parseInt((Number(timeInSeconds))/3600);
  let minutes = parseInt(Number(timeInSeconds)/60)-(hours*60);
  let seconds = (Number(timeInSeconds)%60);
  let formattedSplit = "--:--.--";


  if (timeInSeconds < 60) {
    formattedSplit = seconds.toFixed(2);
  }
  else if (timeInSeconds < 60*60) {

    if (seconds < 10) {
      formattedSplit = minutes + ":0" + seconds.toFixed(2);
    } else {
      formattedSplit = minutes + ":" + seconds.toFixed(2);
    }
  }
  else {

    formattedSplit = hours;

    if (minutes < 10) {
      formattedSplit = formattedSplit + ":0" + minutes;
    } else {
      formattedSplit = formattedSplit + ":" + minutes;
    }

    if (seconds < 10) {
      formattedSplit = formattedSplit + ":0" + seconds.toFixed(2);
    } else {
      formattedSplit = formattedSplit + ":" + seconds.toFixed(2);
    }
  }

  return formattedSplit;
}

export function timesRun(race, desiredSplitVar) {
  let numberTimesRun = "--";
  if (!isNaN(race) && !isNaN(desiredSplitVar)) {
    numberTimesRun = parseInt((Number(race)/Number(desiredSplitVar)));
  }
  return numberTimesRun;
}

// main display component
// calculates splits by dividing total race time by the ratio of
// the desired split to the given race distance
// returns main calculation with table underneath
const SplitsResults = ({ race, desiredSplitVar, hours, minutes, seconds}) => {

  let currentRaceTimeVar = Number(hours)*3600 + Number(minutes)*60 + Number(seconds);

  let mySplit = ((Number(desiredSplitVar)/Number(race))*(Number(currentRaceTimeVar)));

  const events = [
    {
      id: 0,
      event: '400m',
      time: formatResult((Number(400)/Number(race))*(Number(currentRaceTimeVar))),
    },
    {
      id: 1,
      event: '600m',
      time: formatResult((Number(600)/Number(race))*(Number(currentRaceTimeVar))),
    },
    {
      id: 2,
      event: '800m',
      time: formatResult((Number(800)/Number(race))*(Number(currentRaceTimeVar))),
    },
    {
      id: 3,
      event: '1000m',
      time: formatResult((Number(1000)/Number(race))*(Number(currentRaceTimeVar))),
    },
    {
      id: 4,
      event: '1200m',
      time: formatResult((Number(1200)/Number(race))*(Number(currentRaceTimeVar))),
    },
    {
      id: 5,
      event: '1600m',
      time: formatResult((Number(1600)/Number(race))*(Number(currentRaceTimeVar))),
    },
    {
      id: 6,
      event: '1609m (mile)',
      time: formatResult((Number(1609)/Number(race))*(Number(currentRaceTimeVar))),
    },
    {
      id: 7,
      event: '3000m',
      time: formatResult((Number(3000)/Number(race))*(Number(currentRaceTimeVar))),
    },
    {
      id: 8,
      event: '3200m',
      time: formatResult((Number(3200)/Number(race))*(Number(currentRaceTimeVar))),
    },
    {
      id: 9,
      event: '5000m',
      time: formatResult((Number(5000)/Number(race))*(Number(currentRaceTimeVar))),
    },
    {
      id: 10,
      event: '10000m',
      time: formatResult((Number(10000)/Number(race))*(Number(currentRaceTimeVar))),
    }
  ]

return (
  <View>
    <Text style = {styles.mainResult}> { formatResult(mySplit) } </Text>
    <Text style = {{textAlign: 'center', marginBottom: 10}}> For {desiredSplitVar}m ({ timesRun(race, desiredSplitVar) } times) </Text>
    <Text style = {{textAlign: 'center', marginBottom: 30}}> { (formatResult(currentRaceTimeVar - (mySplit * parseInt(Number(race)/Number(desiredSplitVar))))) } for remaining {race - (parseInt(race/desiredSplitVar) * desiredSplitVar)}m </Text>
    <ScrollView style = {{borderWidth: 0.5, borderBottomWidth: 0, borderColor: 'grey'}}>
    {events.map((item, index) => (
        <View key={item.id} style={styles.container}>
          <Text style={styles.text}>  {item.event}</Text>
          <Text style={styles.text}>{item.time}  </Text>
        </View>
      ))}
    </ScrollView>
  </View>
)



}
export default SplitsResults

const styles = StyleSheet.create ({
  container: {
      padding: 10,
      marginTop: 3,
      flexDirection: 'row',
      justifyContent: 'space-between',
      alignItems: 'flex-end',
      borderBottomColor: 'grey',
      borderBottomWidth: 0.5
   },
   text: {
      color: 'black',
      fontSize: 15
   },
   mainResult: {
     width: 250,
     height: 75,
     textAlign: 'center',
     padding: 18,
     fontSize: 30,
     backgroundColor: 'white',
     marginTop: 60,
     marginBottom: 10,
     marginLeft: 60
   }
})
