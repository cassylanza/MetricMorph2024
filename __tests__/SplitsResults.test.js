import React from 'react';
import { render, screen, fireEvent, getByText, toBe, queryAllByText, getByPlaceholderText } from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import { NavigationContainer } from '@react-navigation/native';
import { Actions } from 'react-native-router-flux'; // Mocking react-native-router-flux
import '@testing-library/jest-dom';
import SplitsResults from '../SplitsResults.js';
import { formatResult } from '../SplitsResults.js';

jest.mock('react-native-router-flux', () => ({
  Actions: {
    swimmenu1: jest.fn(),
    trackmenu1: jest.fn(),
    convperfinfoentry: jest.fn(),
    conversionresults: jest.fn(),
    splitsinfoentry: jest.fn(),
    splitsresults: jest.fn(),
    effortinfoentry: jest.fn(),
    effortresults: jest.fn(),
  },
}));

describe('SplitsResults', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<SplitsResults />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('SplitsResults component', () => {
  it('displays the main result time correctly', () => {
    const race = 1600;
    const hours = 0;
    const minutes = 5;
    const seconds = 30;
    const desiredSplitVar = 400;

    const { getByText } = render(
      <SplitsResults race={race} hours={hours} minutes={minutes} seconds={seconds} desiredSplitVar={desiredSplitVar} />
    );

    const currentRaceTimeVar = hours*3600 + minutes*60 + seconds;
    const mySplit = desiredSplitVar/race*(currentRaceTimeVar);
    const expectedMainResult = formatResult(mySplit);
    const elements = screen.queryAllByText(expectedMainResult);

    expect(elements.length).toBeGreaterThan(0);
  });
});

describe('SplitsResults component', () => {
  it('displays the distance clarifier correctly', () => {
    const { getByText } = render(
      <SplitsResults race={1600} hours={0} minutes={6} seconds={0} desiredSplitVar={600} />
    );

    const distanceClarifier1 = getByText(' For 600m (2 times) ');
    const distanceClarifier2 = getByText(' 1:30.00 for remaining 400m ');

    expect(distanceClarifier1).toBeTruthy();
    expect(distanceClarifier2).toBeTruthy();
  });
});

  it('table appears', () => {
    const { getByText } = render(<SplitsResults />);
    expect(getByText("400m")).toBeTruthy();
  });


  describe('formatResult function', () => {
    it('returns "hh:mm:ss" for invalid input', () => {
      expect(formatResult(undefined)).toBe('hh:mm:ss');
      expect(formatResult(null)).toBe('hh:mm:ss');
      expect(formatResult('abc')).toBe('hh:mm:ss');
    });

    it('returns "hh:mm:ss" if timeInSeconds is out of range', () => {
      expect(formatResult(-1)).toBe('hh:mm:ss');
      expect(formatResult(362400)).toBe('hh:mm:ss');
    });

    it('correctly formats time less than 60 seconds', () => {
      expect(formatResult(30)).toBe('30.00');
      expect(formatResult(59.99)).toBe('59.99');
    });

    it('correctly formats time between 60 seconds and 1 hour', () => {
      expect(formatResult(65)).toBe('1:05.00');
      expect(formatResult(3600)).toBe('1:00:00.00');
    });

    it('correctly formats time greater than 1 hour', () => {
      expect(formatResult(3665)).toBe('1:01:05.00');
      expect(formatResult(3666)).toBe('1:01:06.00');
      expect(formatResult(4200)).toBe('1:10:00.00');
      expect(formatResult(4210.23)).toBe('1:10:10.23')
      expect(formatResult(7200)).toBe('2:00:00.00');
    });
  });


})
