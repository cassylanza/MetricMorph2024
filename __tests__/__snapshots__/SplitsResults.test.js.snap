// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SplitsResults renders correctly 1`] = `
<View>
  <Text
    style={
      {
        "backgroundColor": "white",
        "fontSize": 30,
        "height": 75,
        "marginBottom": 10,
        "marginLeft": 60,
        "marginTop": 60,
        "padding": 18,
        "textAlign": "center",
        "width": 250,
      }
    }
  >
     
    hh:mm:ss
     
  </Text>
  <Text
    style={
      {
        "marginBottom": 10,
        "textAlign": "center",
      }
    }
  >
     For 
    m (
    NaN
     times) 
  </Text>
  <Text
    style={
      {
        "marginBottom": 30,
        "textAlign": "center",
      }
    }
  >
     
    hh:mm:ss
     for remaining 
    NaN
    m 
  </Text>
  <RCTScrollView
    style={
      {
        "borderBottomWidth": 0,
        "borderColor": "grey",
        "borderWidth": 0.5,
      }
    }
  >
    <View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderBottomColor": "grey",
            "borderBottomWidth": 0.5,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "black",
              "fontSize": 15,
            }
          }
        >
            
          400m
        </Text>
        <Text
          style={
            {
              "color": "black",
              "fontSize": 15,
            }
          }
        >
          hh:mm:ss
            
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderBottomColor": "grey",
            "borderBottomWidth": 0.5,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "black",
              "fontSize": 15,
            }
          }
        >
            
          600m
        </Text>
        <Text
          style={
            {
              "color": "black",
              "fontSize": 15,
            }
          }
        >
          hh:mm:ss
            
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderBottomColor": "grey",
            "borderBottomWidth": 0.5,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "black",
              "fontSize": 15,
            }
          }
        >
            
          800m
        </Text>
        <Text
          style={
            {
              "color": "black",
              "fontSize": 15,
            }
          }
        >
          hh:mm:ss
            
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderBottomColor": "grey",
            "borderBottomWidth": 0.5,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "black",
              "fontSize": 15,
            }
          }
        >
            
          1000m
        </Text>
        <Text
          style={
            {
              "color": "black",
              "fontSize": 15,
            }
          }
        >
          hh:mm:ss
            
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderBottomColor": "grey",
            "borderBottomWidth": 0.5,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "black",
              "fontSize": 15,
            }
          }
        >
            
          1200m
        </Text>
        <Text
          style={
            {
              "color": "black",
              "fontSize": 15,
            }
          }
        >
          hh:mm:ss
            
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderBottomColor": "grey",
            "borderBottomWidth": 0.5,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "black",
              "fontSize": 15,
            }
          }
        >
            
          1600m
        </Text>
        <Text
          style={
            {
              "color": "black",
              "fontSize": 15,
            }
          }
        >
          hh:mm:ss
            
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderBottomColor": "grey",
            "borderBottomWidth": 0.5,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "black",
              "fontSize": 15,
            }
          }
        >
            
          1609m (mile)
        </Text>
        <Text
          style={
            {
              "color": "black",
              "fontSize": 15,
            }
          }
        >
          hh:mm:ss
            
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderBottomColor": "grey",
            "borderBottomWidth": 0.5,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "black",
              "fontSize": 15,
            }
          }
        >
            
          3000m
        </Text>
        <Text
          style={
            {
              "color": "black",
              "fontSize": 15,
            }
          }
        >
          hh:mm:ss
            
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderBottomColor": "grey",
            "borderBottomWidth": 0.5,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "black",
              "fontSize": 15,
            }
          }
        >
            
          3200m
        </Text>
        <Text
          style={
            {
              "color": "black",
              "fontSize": 15,
            }
          }
        >
          hh:mm:ss
            
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderBottomColor": "grey",
            "borderBottomWidth": 0.5,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "black",
              "fontSize": 15,
            }
          }
        >
            
          5000m
        </Text>
        <Text
          style={
            {
              "color": "black",
              "fontSize": 15,
            }
          }
        >
          hh:mm:ss
            
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderBottomColor": "grey",
            "borderBottomWidth": 0.5,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "black",
              "fontSize": 15,
            }
          }
        >
            
          10000m
        </Text>
        <Text
          style={
            {
              "color": "black",
              "fontSize": 15,
            }
          }
        >
          hh:mm:ss
            
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
